Fix Select scroll to selected option

TomasBoda opened the pull request:
## Description

Important changes:
- when the options are being loaded, the select is disabled
- when user selects an option, upon opening the menu again, the menu scrolls to the selected option group header
- when the menu is open, options are cached
- when the menu is closed, options can be updated

## Related Issue

Resolve #18749 

## Screenshots:

https://github.com/user-attachments/assets/90f6b895-df63-46f8-bc21-f756b7f003a2

tomasklim commented on packages/suite/src/views/wallet/trading/common/TradingForm/TradingFormInput/TradingFormInputAccount.tsx:
```diff
@@ -37,6 +38,11 @@ export const TradingFormInputAccount = <
     const optionGroups = useTradingBuildAccountGroups(type);
 
+    const exchangeInfo = useSelector(state => state.wallet.tradingNew.exchange.exchangeInfo);
+    const sellInfo = useSelector(state => state.wallet.tradingNew.sell.sellInfo);
```
there should be `isLoading` in `tradingNew`. Can it be used or not?

tomasklim replied:
because in case fetch fails, it will be stuck in loading state

adderpositive replied:
For loading info from the API (common info, buyInfo, sellInfo, exchangeInfo) serves `isLoading` at the top of the trading root. 
You can use the `selectTradingLoadingAndTimestamp` selector for that.

There is also `isLoading` in the sub-reducers (buy, sell, exchange) used for loading quotes.

TomasBoda replied:
yes it can, changed it

tomasklim commented on suite-common/trading/src/thunks/common/__tests__/loadInitialDataThunk.test.ts:
```diff
@@ -213,51 +181,6 @@ describe('loadInitialDataThunk', () => {
         await testUpdatedInfoData('outdated');
     });
 
-    it('should keep same version of data without update', async () => {
```
are there test covering the logic after changes?

TomasBoda replied:
added tests for the separate loading of buy/sell/exchange sections

TomasBoda replied:
I reverted the separate loading of sections, because it caused issues in other parts of Suite. Now the select works even though they are loaded all at once, so the tests stay the same as they were

adderpositive commented on suite-common/trading/src/thunks/common/loadInitialDataThunk.ts:
```diff
@@ -42,7 +42,7 @@ export const loadInitialDataThunk = createThunk(
         dispatch(tradingActions.setTradingActiveSection(activeSection));
 
-        if (account && !isLoading && (isDifferentAccount || areDataOutdated)) {
+        if (account && !isLoading) {
```
`invityAPI.getInfo()` will be called every time, regardless of whether the data is outdated or is for a different account.

adderpositive commented on suite-common/trading/src/thunks/common/loadInitialDataThunk.ts:
```diff
@@ -58,19 +58,22 @@ export const loadInitialDataThunk = createThunk(
             }
 
-            if (isDifferentAccount || !buyInfo) {
+            if (activeSection === 'buy' && (isDifferentAccount || !buyInfo || areDataOutdated)) {
```
`activeSection === 'buy'` should not be here. I got your point, but it will not work for another part of the suite, when we need to get data regardless of the `activeSection`, for example, `TokenList` page - `Swap` will be unavailable unless you visit Swap in the trading before.

![image](https://github.com/user-attachments/assets/ffb293d3-7296-4461-a744-9db924dd9738)

TomasBoda replied:
you're right, thanks for the catch. It doesn't make sense to load `activeSection` only then, I reverted it back to as it was before

adderpositive commented on packages/components/src/components/form/Select/customComponents.tsx:
```diff
@@ -22,23 +25,37 @@ export const Control = ({ 'data-testid': dataTest, ...controlProps }: ControlCom
+}: OptionComponentProps) => {
+    const ref = useRef<HTMLDivElement>();
+
+    useEffect(() => {
```
`useDeepCompareEffect` should solve the problem with creating a new `deepEqual` function.

TomasBoda replied:
It doesn't work, it scrolls to the very bottom of the dropdown everytime for some reason. I kept the `deepEqual`

adderpositive replied:
Maybe `useDeepCompareEffect` is for a bit different use case. Sorry for misleading.

adderpositive commented on packages/suite/src/views/wallet/trading/common/TradingForm/TradingFormInput/TradingFormInputAccount.tsx:
```diff
@@ -37,6 +38,11 @@ export const TradingFormInputAccount = <
     const optionGroups = useTradingBuildAccountGroups(type);
 
+    const exchangeInfo = useSelector(state => state.wallet.tradingNew.exchange.exchangeInfo);
+    const sellInfo = useSelector(state => state.wallet.tradingNew.sell.sellInfo);
```
For `exchangeInfo`, `sellInfo`, and `buyInfo` are prepared selectors, which transform values correctly.

`selectTradingExchangeInfo`
`selectTradingSellInfo`
`selectTradingBuyInfo`

Please use them if they are necessary.

TomasBoda replied:
I am now using the `isLoading` state for this

adderpositive commented on packages/components/src/components/form/Select/Select.tsx:
```diff
@@ -346,34 +314,14 @@ export const Select = ({
 
-        return rest.options?.map((option, index) => ({ ...option, [OPTION_INDEX_NAME]: index }));
-    }, [rest.options, isGrouped]);
+    const [selectedOption, setSelectedOption] = useState<any>(rest.value);
```
`any`? Can we use `Option | undefined` or something like that?

TomasBoda replied:
done

adderpositive reviewed (changes requested):
1. UX - It would be nice to also see under what account it belongs to. What do you think? But I am not sure if there will be, for example, 20 coins under one account…
![image](https://github.com/user-attachments/assets/f88de07e-5075-436e-b451-b91881a28544)

2. Sometimes the scrolling is jumping, check the video.
https://github.com/user-attachments/assets/83b2e468-3410-400c-9bc9-350f883203be

TomasBoda commented:
> 1. UX - It would be nice to also see under what account it belongs to. What do you think? But I am not sure if there will be, for example, 20 coins under one account…
>    ![image](https://private-user-images.githubusercontent.com/15160619/452955020-f88de07e-5075-436e-b451-b91881a28544.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDk0NjUwODksIm5iZiI6MTc0OTQ2NDc4OSwicGF0aCI6Ii8xNTE2MDYxOS80NTI5NTUwMjAtZjg4ZGUwN2UtNTA3NS00MzZlLWI0NTEtYjkxODgxYTI4NTQ0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjA5VDEwMjYyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkwNWVmNzlhOWM3NjVlMzVjNGQ2MDdiOTc1YTg5NjJkYjNiYmUzYWJkMmEyZGY4MGM3NTg3OGU0M2UwNDdiZmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.szgJMu-MLqhNk1Ern_F60-Lq8XSKydAk_BEEnBODo2U)
> 2. Sometimes the scrolling is jumping, check the video.
>    https://github.com/user-attachments/assets/83b2e468-3410-400c-9bc9-350f883203be
> 3. I don't understand changes in trading. 🤷‍♂️ Why the changes?

1. This was my intention from the start, but as token info is loading for a few tens of seconds after the select is ready, it refreshes the options and makes it difficult to handle scrolling in the select. The current solution is the only one that works at least somehow consistently
2. I fixed the jumping to some extent
3. The changes were intended to load the sections separately, but as you pointed out, it causes problems, so I reverted it for the time being

adderpositive commented on packages/components/src/components/form/Select/customComponents.tsx:
```diff
@@ -22,23 +25,38 @@ export const Control = ({ 'data-testid': dataTest, ...controlProps }: ControlCom
+}: OptionComponentProps) => {
+    const ref = useRef<HTMLDivElement>();
+
+    useEffect(() => {
```
Replaced with `useDeepCompareEffect`, it should prevent jumping in the select list

adderpositive reviewed (changes requested):
Thank you for the updates. Select list still jumps a bit.

tomasklim commented:
/rebase
